### DISCUSSION TITLE
riverpod_generator for StateNotifier

### DISCUSSION THREAD

**Discussion by derolf** [rrousselGit/riverpod#2253] [Q&A]
Is there a way to use `riverpod_generator` to generate a `StateNotifierProvider` for a function returning a `StateNotifier<T>`?

**Comment by rrousselGit** [comment]
No.
The expected solution is to use:

```dart
@riverpod
class YourNotifier extends _$YourNotifier {
  @override
  YourState build() => YourState();
}
```

**Reply by fredgrott** [reply]
Thank you @rrousselGit as answer my use case as well.

**Reply by fa0311** [reply]
Can you add a feature that would make it simpler to write?
For example:

```dart
@Riverpod(notifier: true)
int counter(CounterRef ref) => 0;
```

**Reply by rrousselGit** [reply]
I have no plan to support StateProvider, no.

**Comment by ValentinVignal** [comment]
There is a package to generated the equivalent of the `StateProvider`: 

https://pub.dev/packages/riverpod_state_provider_annotation
https://pub.dev/packages/riverpod_state_provider_generator

```dart
# main.dart
import 'package:riverpod_state_provider_annotation/riverpod_state_provider_annotation.dart';

part 'main.g.dart';
part 'main.rsp.dart';

@RiverpodStateProvider()
int myState() {
  return 0;
}

ref.watch(myStateProvider);
ref.read(myStateProvider.notifier).state++;
ref.read(myStateProvider.notifier).update((state) => state + 1);

ProviderScope(
  overrides: [
    myStateProvider.overrideWithValue(10),
  ],
)
```

It still has some limitations but for simple usages (usually the case for `StateProvider`), it does the job
